contracts/btc-staking/src/validation/mod.rs

```diff
@@ -241,3 +255,107 @@ pub fn verify_active_delegation(
 
     Ok(())
 }
+
+pub fn verify_undelegation(
+    params: &Params,
+    btc_del: &BtcDelegation,
+    sig: &Binary,
+) -> Result<(), ContractError> {
+    // The following code is marked with `#[cfg(feature = "full-validation")]`
+    // so that it is included in the build if the `full-validation` feature is
+    // enabled.
+    // TODO: fix contract size when full-validation is enabled
+    #[cfg(feature = "full-validation")]
+    {
+        /*
+            Verify the signature on the unbonding tx is from the delegator
+        */
+
+        // get keys
+        let (staker_pk, fp_pks, cov_pks) = get_pks(
+            btc_del.btc_pk_hex.clone(),
+            btc_del.fp_btc_pk_list.clone(),
+            params.covenant_pks.clone(),
+        )?;
+
+        // get the unbonding path script
+        let staking_tx: Transaction = deserialize(&btc_del.staking_tx)
+            .map_err(|_| ContractError::InvalidBtcTx(btc_del.staking_tx.encode_hex()))?;
+        let staking_output = &staking_tx.output[btc_del.staking_output_idx as usize];
+        let staking_time = (btc_del.end_height - btc_del.start_height) as u16;
+        let babylon_script_paths = babylon_btcstaking::scripts_utils::BabylonScriptPaths::new(
+            &staker_pk,
+            &fp_pks,
+            &cov_pks,
+            params.covenant_quorum as usize,
+            staking_time,
+        )?;
+        let unbonding_path_script = babylon_script_paths.unbonding_path_script;
+
+        // get unbonding tx
+        let unbonding_tx: Transaction = deserialize(&btc_del.undelegation_info.unbonding_tx)
+            .map_err(|_| {
+                ContractError::InvalidBtcTx(btc_del.undelegation_info.unbonding_tx.encode_hex())
+            })?;
+
+        // get the staker's signature on the unbonding tx
+        let staker_sig = k256::schnorr::Signature::try_from(sig.as_slice())
+            .map_err(|e| ContractError::SecP256K1Error(e.to_string()))?;
+
+        // Verify the signature
+        babylon_btcstaking::sig_verify::verify_transaction_sig_with_output(
+            &unbonding_tx,
+            staking_output,
+            unbonding_path_script.as_script(),
+            &staker_pk,
+            &staker_sig,
+        )
+        .map_err(|e| ContractError::SecP256K1Error(e.to_string()))?;
+    }
+
+    // make static analyser happy with unused parameters
+    #[cfg(not(feature = "full-validation"))]
+    let _ = (params, btc_del, sig);
+
+    Ok(())
+}
+
+pub fn verify_slashed_delegation(
+    active_delegation: &BtcDelegation,
+    slashed_fp_sk_hex: String,
```

Comment:
Here as well, it's in general better to use 
```suggestion
    slashed_fp_sk_hex: &str,
```